autometa/config/databases.py

```diff
@@ -66,230 +97,504 @@ class Databases:
+
+        Returns
+        -------
+        NoneType
+
+        """
+        hmms = (os.path.join(self.markers_dir, fn) for fn in os.listdir(self.markers_dir) if fn.endswith('.hmm'))
+        hmm_search_str = os.path.join('autometa/databases/markers/','*.h3*')
+        pressed_hmms = glob(hmm_search_str)
+        pressed_hmms = set(os.path.realpath(os.path.splitext(fp)[0]) for fp in pressed_hmms)
+        for fp in hmms:
+            if fp in pressed_hmms:
+                continue
+            hmmer.hmmpress(fp)
+
+    def download_markers(self, options):
+        """Download markers database files and amend user config to reflect this.
 
         Parameters
         ----------
-        options : set
-            Description of parameter `options`.
+        options : iterable
+            iterable containing options in 'markers' section to download.
 
         Returns
         -------
         NoneType
+            Will update provided `options` in `self.config`.
 
         Raises
         -------
         ConnectionError
             marker file download failed.
 
         """
-        if not os.path.exists(self.markers_dir):
-            os.makedirs(self.markers_dir)
         for option in options:
-            url = DEFAULT_CONFIG.get('database_urls', option)
+            # First retrieve the markers file url from `option` in `markers` section
+            url = self.config.get('database_urls', option)
             if self.config.has_option('markers', option):
                 outfpath = self.config.get('markers', option)
             else:
                 outfname = os.path.basename(url)
                 outfpath = os.path.join(self.markers_dir, outfname)
+
             if self.dryrun:
                 logger.debug(f'UPDATE: (markers,{option}): {outfpath}')
                 self.config.set('markers', option, outfpath)
                 continue
-            with requests.Session() as session:
+
+            # Retrieve markers file and write contents to `outfpath`
+            with requests.Session() as session, open(outfpath, 'w') as fh:
                 resp = session.get(url)
-            if not resp.ok:
-                raise ConnectionError(f'Failed to retrieve {url}')
-            with open(outfpath, 'w') as outfh:
-                outfh.write(resp.text)
+                if not resp.ok:
+                    raise ConnectionError(f'Failed to retrieve {url}')
+                fh.write(resp.text)
             self.config.set('markers', option, outfpath)
-            if outfpath.endswith('.hmm'):
-                hmmer.hmmpress(outfpath)
+        self.press_hmms()
+
+    def get_missing(self, section=None, validate=False):
+        """Get all missing database files in `options` from `sections`
+        in config.
+
+        If the database file already exists, an md5 checksum comparison will be
+        performed bewteen the current and file's remote md5 to ensure file integrity
+        prior to checking the respective file as valid.
```

Comment:
This function doesn't seem to do this though. If the file exists then it returns `True`. So we need to actually add the md5 comparison. Also, as written right now I think this will fail for the marker files at least because an md5 file is not made then they are downloaded (see other comment).

Reply:
I'm going to try returning an empty `dict` instead mixed dtypes (bool and dict) to try to make this more readable.

Reply:
Right now though,

if a user were to specify `validate = True`, then the `if validate: return True` block will _only_ be encountered if `if os.path.exists(fpath): continue` fails. Meaning there is a missing file, so return `True`. Otherwise, if `validate = True`, and no missing files are encountered, the function will return `False`.